cowbird/services/impl/geoserver.py

```diff
@@ -38,156 +81,174 @@ def __init__(self, settings, name, **kwargs):
         """
         super(Geoserver, self).__init__(settings, name, **kwargs)
         self.api_url = "{}/rest".format(self.url)
-        self.auth = (self.admin_user, self.admin_password)
         self.headers = {"Content-type": CONTENT_TYPE_JSON}
+        self.admin_user = kwargs.get(SERVICE_ADMIN_USER, None)
+        self.admin_password = kwargs.get(SERVICE_ADMIN_PASSWORD, None)
+        self.auth = (self.admin_user, self.admin_password)
+
+    #
+    # Implementation of parent classes' functions
+    #
 
+    # Service class functions
     def get_resource_id(self, resource_full_name):
-        # type (str) -> str
+        # type:(str) -> str
         raise NotImplementedError
 
     def user_created(self, user_name):
-        res = chain(create_workspace.si(user_name) | create_datastore.si(user_name))
+        self._create_datastore_dir(user_name)
+        res = chain(create_workspace.si(user_name), create_datastore.si(user_name))
         res.delay()
+        LOGGER.info("Start monitoring datastore of created user [%s]", user_name)
+        Monitoring().register(self._shapefile_folder_dir(user_name), True, Geoserver)
 
     def user_deleted(self, user_name):
         remove_workspace.delay(user_name)
+        LOGGER.info("Stop monitoring datastore of created user [%s]", user_name)
+        Monitoring().unregister(self._shapefile_folder_dir(user_name), self)
 
     def permission_created(self, permission):
         raise NotImplementedError
 
     def permission_deleted(self, permission):
         raise NotImplementedError
 
+    # FSMonitor class functions
+    @staticmethod
+    def get_instance():
+        """
+        Return the Geoserver singleton instance from the class name used to retrieve the FSMonitor from the DB.
+        """
+        return ServiceFactory().get_service("Geoserver")
+
+    def on_created(self, filename):
+        """
+        Call when a new file is found.
+
+        :param filename: Relative filename of a new file
+        """
+        if filename.endswith(".shp"):
+            workspace_name, shapefile_name = self._get_shapefile_info(filename)
+            # Small wait time to prevent unnecessary failure since shapefile is a multi file format
+            sleep(1)
```

Comment:
Does this hang the whole monitoring thread for 10secs if I drop 10 .shp files?

Reply:
Maybe the validate_shapefile function could wait for 1 sec before doing its job? This way there is no sleep on the main thread.

Reply:
Good point, I'll move the sleep to function